src/responsibleai/rai_analyse/_rai_insight_data.py

```diff
@@ -0,0 +1,475 @@
+                leaf = {
+                    "n_samples": node.n_samples,
+                    "treatment": node.treatment,
+                    "parents": parents
+                }
+
+                leaves_collection.append(leaf)
+
+            else:
+                this_node = {
+                    "feature": node.feature,
+                    "right_comparison": node.right_comparison,
+                    "comparison_value": node.comparison_value
+                }
+
+                # dictionary assignment is by reference. Need deepcopy for by value assignment
+                node_left = copy.deepcopy(this_node)
+                node_right = copy.deepcopy(this_node)
+
+                node_left["path"] = "left"
+                node_right["path"] = "right"
+                pat_left = parents + [node_left]
+                pat_right = parents + [node_right]
+
+                visit(node.left, pat_left, leaves_collection)
+                visit(node.right, pat_right, leaves_collection)
+
+            return leaves_collection
+
+        ca = self.raiinsight.causal.get_data()[0]
+
+        return {
+            "global_effect": {k["feature"]: k for k in ca.global_effects},
+            "policy_treatments": {k.treatment_feature: visit(k.policy_tree, [], []) for k in ca.policies},
+            "top_local_policies": {k.treatment_feature: k.local_policies[:3] for k in ca.policies}
+        }
+
+    def to_tree_map(self, tree):
+        return {x["id"]: x for x in tree}
+
+    def get_filter_conditions(self, tree_map, nodeid):
+        filter_conditions = []
+        current_index = nodeid
+        while current_index in tree_map.keys():
+            current_node = tree_map[current_index]
+            if current_node["method"]:
+                filter_conditions.append(current_node["condition"])
+            current_index = current_node["parentId"]
+        return filter_conditions
+
+    def get_min_max_nodes(self, treemap, n):
+        node_list = [{"id": v["id"], "metricValue": v["metricValue"]} for k, v in treemap.items()]
+        sorted_nodes = sorted(node_list, key=lambda d: d['metricValue'])
+        node_size = len(sorted_nodes)
+        if node_size < 2*n:
+            mid_point = int(node_size/2)
+            return sorted_nodes[0:mid_point], sorted_nodes[mid_point:node_size]
+        return sorted_nodes[:n], sorted_nodes[-n:]
+
+    def get_feature_statistics(self, feature, dataset=None):
+        if dataset is None:
+            dataset = self.raiinsight.test
+        return dataset[feature].value_counts()
+
+    def get_cohort_data(self, filtermap):
+        filtered_dataset = self.raiinsight.test[filtermap]
+        filtered_y_pred = self.get_y_pred()[filtermap]
+        filtered_y_test = self.get_y_test()[filtermap]
+        return {
+            "y_pred": filtered_y_pred,
+            "y_test": filtered_y_test,
+            "population": len(filtered_dataset) / len(self.raiinsight.test)
+        }
+
+
+class PdfDataGen:
```

Comment:
Is it possible to put the classes in their own individual files? GitHub doesn't want to display this file by default because it's so long.

Reply:
okay - will separate them into three => core_components, regression_components and classification_components.

Reply:
done!